fix: Add support for asset variations and rental options to catalogs

kirre-bylund opened the pull request:
Note that the changes are done for both v1 and v2 catalog listing.

- Catalog Response has a dict of assets so needs a dict of asset variations as well
  - Solution: Parse assets but place optionals in separate dict
- Catalog Entries of type Group has a list of Associations with just the basic key
  - Failed to solve: Figure out a clean way to get the variations here. Since the associations hold just the basic key (listing id and item id) there's no way to figure out cleanly which association goes with which variation without backend changes.
- Inlined Catalog Entries has a single asset_details var
  - Solution: Add optionals array and populate with all asset variants that match the basic key if there's not a single asset detail that matches. Populate optionals list only if entry asset details has no single match.
- Inlined Group has a list of assetDetails
  - Solution: Add single match if found for the association, otherwise add all optionals matching basic key BUT keep track of previously added basic keys so we don't add the same optionals multiple times for multiple associations with the same basic key.

JohannesLoot reviewed (approved):
All good!
Merge away!

kirre-bylund commented on LootLockerSDK/Source/LootLockerSDK/Private/GameAPI/LootLockerCatalogRequestHandler.cpp:
```diff
@@ -85,6 +96,17 @@ FLootLockerInlinedCatalogEntry::FLootLockerInlinedCatalogEntry(const FLootLocker
+					// Fallback for older versions where Asset Variations and Rental Options existed
+					for (const auto& Pair : CatalogListing.Asset_Details)
+					{
+						if (Pair.Key.Catalog_listing_id == DetailsKey.Catalog_listing_id && Pair.Key.Item_id == DetailsKey.Item_id)
```
Fixed

kirre-bylund commented on LootLockerSDK/Source/LootLockerSDK/Private/GameAPI/LootLockerCatalogRequestHandler.cpp:
```diff
@@ -208,6 +241,17 @@ FLootLockerInlinedCatalogEntry::FLootLockerInlinedCatalogEntry(const FLootLocker
+					// Fallback for older versions where Asset Variations and Rental Options existed
+					for (const auto& Pair : CatalogListing.Asset_Details)
+					{
+						if (Pair.Key.Catalog_listing_id == DetailsKey.Catalog_listing_id && Pair.Key.Item_id == DetailsKey.Item_id)
```
Fixed

kirre-bylund commented on LootLockerSDK/Source/LootLockerSDK/Private/GameAPI/LootLockerCatalogRequestHandler.cpp:
```diff
@@ -85,6 +96,17 @@ FLootLockerInlinedCatalogEntry::FLootLockerInlinedCatalogEntry(const FLootLocker
+					{
+						if (Pair.Key.Catalog_listing_id == DetailsKey.Catalog_listing_id && Pair.Key.Item_id == DetailsKey.Item_id)
+						{
+							this->OptionalAssetDetailVariants.Add(Pair.Value);
```
Fixed

kirre-bylund commented on LootLockerSDK/Source/LootLockerSDK/Private/GameAPI/LootLockerCatalogRequestHandler.cpp:
```diff
@@ -208,6 +241,17 @@ FLootLockerInlinedCatalogEntry::FLootLockerInlinedCatalogEntry(const FLootLocker
+					{
+						if (Pair.Key.Catalog_listing_id == DetailsKey.Catalog_listing_id && Pair.Key.Item_id == DetailsKey.Item_id)
+						{
+							this->OptionalAssetDetailVariants.Add(Pair.Value);
```
Fixed

kirre-bylund committed:
fix: Add support for asset variations and rental options to catalogs

Note that the changes are done for both v1 and v2 catalog listing.

- Catalog Response has a dict of assets so needs a dict of asset variations as well
  - Solution: Parse assets but place optionals in separate dict
- Catalog Entries of type Group has a list of Associations with just the basic key
  - Failed to solve: Figure out a clean way to get the variations here. Since the associations hold just the basic key (listing id and item id) there's no way to figure out cleanly which association goes with which variation without backend changes.
- Inlined Catalog Entries has a single asset_details var
  - Solution: Add optionals array and populate with all asset variants that match the basic key if there's not a single asset detail that matches. Populate optionals list only if entry asset details has no single match.
- Inlined Group has a list of assetDetails
  - Solution: Add single match if found for the association, otherwise add all optionals matching basic key BUT keep track of previously added basic keys so we don't add the same optionals multiple times for multiple associations with the same basic key.

JohannesLoot reviewed (approved):
Checks out!
Merge away!
